examples/c/ex_storage_source.c

```diff
@@ -0,0 +1,1142 @@
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * ex_storage_source.c
+ * 	demonstrates how to use the custom storage source interface
+ */
+#include <test_util.h>
+
+#ifdef __GNUC__
+#if __GNUC__ > 7 || (__GNUC__ == 7 && __GNUC_MINOR__ > 0)
+/*
+ * !!!
+ * GCC with -Wformat-truncation complains about calls to snprintf in this file.
+ * There's nothing wrong, this makes the warning go away.
+ */
+#pragma GCC diagnostic ignored "-Wformat-truncation"
+#endif
+#endif
+
+/*
+ * This example code uses pthread functions for portable locking, we ignore errors for simplicity.
+ */
+static void
+allocate_storage_source_lock(pthread_rwlock_t *lockp)
+{
+    error_check(pthread_rwlock_init(lockp, NULL));
+}
+
+static void
+destroy_storage_source_lock(pthread_rwlock_t *lockp)
+{
+    error_check(pthread_rwlock_destroy(lockp));
+}
+
+static void
+lock_storage_source(pthread_rwlock_t *lockp)
+{
+    error_check(pthread_rwlock_wrlock(lockp));
+}
+
+static void
+unlock_storage_source(pthread_rwlock_t *lockp)
+{
+    error_check(pthread_rwlock_unlock(lockp));
+}
+
+/*
+ * Example storage source implementation, using memory buffers to represent objects.
+ */
+typedef struct {
+    WT_STORAGE_SOURCE iface;
+
+    /*
+     * WiredTiger performs schema and I/O operations in parallel, all storage sources and file
+     * handle access must be thread-safe. This example uses a single, global storage source lock for
+     * simplicity; real applications might require finer granularity, for example, a single lock for
+     * the storage source handle list and per-handle locks serializing I/O.
+     */
+    pthread_rwlock_t lock; /* Lock */
+
+    int opened_object_count;
+    int opened_unique_object_count;
+    int closed_object_count;
+    int read_ops;
+    int write_ops;
+
+    /* Queue of file handles */
+    TAILQ_HEAD(demo_file_handle_qh, demo_file_handle) fileq;
+
+    WT_EXTENSION_API *wtext; /* Extension functions */
+
+} DEMO_STORAGE_SOURCE;
+
+typedef struct demo_file_handle {
+    WT_FILE_HANDLE iface;
+
+    /*
+     * Add custom file handle fields after the interface.
+     */
+    DEMO_STORAGE_SOURCE *demo_ss; /* Enclosing storage source */
+
+    TAILQ_ENTRY(demo_file_handle) q; /* Queue of handles */
+    uint32_t ref;                    /* Reference count */
+
+    char *buf;      /* In-memory contents */
+    size_t bufsize; /* In-memory buffer size */
+
+    size_t size; /* Read/write data size */
+} DEMO_FILE_HANDLE;
+
+/*
+ * Extension initialization function.
+ */
+#ifdef _WIN32
+/*
+ * Explicitly export this function so it is visible when loading extensions.
+ */
+__declspec(dllexport)
+#endif
+  int demo_storage_source_create(WT_CONNECTION *, WT_CONFIG_ARG *);
+
+/*
+ * Forward function declarations for storage source API implementation
```

Comment:
End the sentence with a period. There are a bunch of comments that don't end in a period. Please look for more. I won't make a comment at every one.